docker_image_sync.sh

```diff
@@ -0,0 +1,342 @@
+      h)
+        Usage;
+        exit 0;
+        ;;
+      *)
+        Usage;
+        Fail "\nInvalid option passed to \"$(basename ${0})\"; exiting. See Usage below.\n";
+        ;;
+    esac
+  done
+  shift $((OPTIND-1));
+}
+
+#################################
+#
+### This section is for functions related to the main execution of the program.
+### Functions in this section perform the following tasks:
+###   - Check to ensure EUID is 0 before attempting sync
+###   - Ensure all required parameters are set before attempting sync
+###   - Ensuring appropriate directories exist for mirror
+###   - Ensuring appropriate dirs/files exist per repo
+###   - Ensuring repo metadata is populated before attempting sync
+###   - Ensuring required sync providers are installed and accesssible
+###   - Performing the sync
+###   - Reporting on the success of the sync
+#
+################################################################################
+
+Validate_variables() {
+
+  # This santizes the directory name of spaces or any other undesired characters.
+  mirror_tld="${mirror_tld//[^a-zA-Z1-9_-/]}";
+  mirror_tld=${mirror_tld//\"};
+  if [[ "${mirror_tld:0:1}" != "/" ]]; then
+    Usage;
+    Fail "\nAbsolute paths only, please; exiting.\n";
+  else
+    # Drops the last "/" from the value of mirror_tld to ensure uniformity for functions using it.
+    # Note: as a side-effect, this effective prevents using just "/" as the value for mirror_tld.
+    mirror_tld="${mirror_tld%\/}";
+  fi
+
+  # Ensures repo_name_array is not empty
+  #if [[ -z "${repo_name_array}" ]]; then
+  #  Fail "Bashellite requires at least one valid repository.";
+  #fi
+
+  # If the mirror_tld is unset or null; then exit.
+  # Since the last "/" was dropped in Parse_parameter,
+  # If user passed "/" for mirror_tld value, it effectively becomes "" (null).
+  if [[ -z "${mirror_tld}" ]]; then
+    Usage;
+    Fail "\nPlease set the desired location of the local mirror; exiting.\n";
+  fi
+}
+
+# This function creates/validates the file/directory framework for the requested repo.
+Validate_repo_framework() {
+  if [[ -n "${repo_name}" ]]; then
+    Info "Creating/validating directory and file structure for mirror and repo (${repo_name})...";
+    #mkdir -p "${providers_tld}";
+    mirror_repo_name="${repo_name//__/\/}";
+    if [[ ! -d "${mirror_tld}" ]]; then
+      Fail "Mirror top-level directory (${mirror_tld}) does not exist!"
+    else
+      mkdir -p "${mirror_tld}/${mirror_repo_name}/" &>/dev/null \
+      || Fail "Unable to create directory (${mirror_tld}/${mirror_repo_name}); check permissions."
+    fi
+  fi
+}
+
+# This function performs the actual sync of the repository
+Sync_repository() {
+  image_name_array=()
+
+  for line in $(cat ${config_file}); do
+    # Check to see if tags are listed
+    tag_index=0
+    tag_index=`expr index "${line}" ':'`
+    tags_found=""
+    tags_found="${line:${tag_index}}"
+    docker_registry_url="index.docker.io"    
```

Comment:
Is this configurable in the config_file? We will have a need to pull from other places like registry.access.redhat.com for OpenShift bits.